clang/lib/Headers/cheriintrin.h

```diff
@@ -0,0 +1,132 @@
+#define cheri_round_representable_length(x)                                    \
+  __builtin_cheri_round_representable_length(x)
+#define cheri_representable_alignment_mask(x)                                  \
+  __builtin_cheri_representable_alignment_mask(x
+#define cheri_setbounds(x, y) __builtin_cheri_bounds_set((x), (y))
+#define cheri_setbounds_exact(x, y) __builtin_cheri_bounds_set_exact((x), (y))
+
+/* Object types, sealing and unsealing: */
+typedef long cheri_otype_t;
+#if defined(__mips__) || defined(__riscv)
+/* CHERI-MIPS and CHERI-RISCV use negative numbers for hardware-interpreted
+ * otypes */
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)-1)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)-2)
+#else
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)0)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)1)
+#endif
+#define cheri_gettype(x) __builtin_cheri_type_get(x)
+#define cheri_is_sealed(x) __builtin_cheri_sealed_get(x)
+#define cheri_is_sentry(x) (__builtin_cheri_type_get(x) == CHERI_OTYPE_SENTRY)
+#define cheri_is_unsealed(x) (!__builtin_cheri_sealed_get(x))
+/* TODO: builtins for indirect sentries */
+#define cheri_seal_entry(x) __builtin_cheri_seal_entry(x)
+#define cheri_seal(x, y) __builtin_cheri_seal((x), (y))
+#define cheri_unseal(x, y) __builtin_cheri_unseal((x), (y))
+
+/* Reconstruct capabilities from raw data: */
+#define cheri_build_cap(x, y) __builtin_cheri_cap_build((x), (y))
+#define cheri_conditional_seal(x, y) __builtin_cheri_conditional_seal((x), (y))
+#define cheri_copy_type(x, y) __builtin_cheri_cap_type_copy((x), (y))
+
+/* Capability permissions: */
+enum __attribute__((flag_enum, enum_extensibility(open))) CheriPermissions {
+  CHERI_PERM_GLOBAL = __CHERI_CAP_PERMISSION_GLOBAL__,
+  CHERI_PERM_EXECUTE = __CHERI_CAP_PERMISSION_PERMIT_EXECUTE__,
+  CHERI_PERM_LOAD = __CHERI_CAP_PERMISSION_PERMIT_LOAD__,
+  CHERI_PERM_STORE = __CHERI_CAP_PERMISSION_PERMIT_STORE__,
+  CHERI_PERM_LOAD_CAP = __CHERI_CAP_PERMISSION_PERMIT_LOAD_CAPABILITY__,
+  CHERI_PERM_STORE_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_CAPABILITY__,
+  CHERI_PERM_STORE_LOCAL_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_LOCAL__,
+  CHERI_PERM_SEAL = __CHERI_CAP_PERMISSION_PERMIT_SEAL__,
+  CHERI_PERM_CCALL = __CHERI_CAP_PERMISSION_PERMIT_CCALL__,
+  CHERI_PERM_UNSEAL = __CHERI_CAP_PERMISSION_PERMIT_UNSEAL__,
+  CHERI_PERM_SYSTEM_REGS = __CHERI_CAP_PERMISSION_ACCESS_SYSTEM_REGISTERS__,
+  /* TODO: architecture-dependent permissions */
+};
+typedef enum CheriPermissions cheri_perms_t;
+#define cheri_getperms(x) ((cheri_perms_t)__builtin_cheri_perms_get(x))
+#define cheri_andperms(x, y) __builtin_cheri_perms_and((x), (cheri_perms_t)(y))
+#define cheri_clearperms(x, y) cheri_andperms((x), ~(y))
+
+/* Accessors for capability registers. Currently exposes DDC,PCC and stack. */
+#define cheri_getddc() __builtin_cheri_global_data_get()
+#define cheri_getpcc() __builtin_cheri_program_counter_get()
+#define cheri_getstack() __builtin_cheri_stack_get()
```

Comment:
I thought `csp` got moved into a perfectly boring gp cap reg?

Reply:
It is, but there is no easy way to read it: https://cheri-compiler-explorer.cl.cam.ac.uk/z/jPKXwP

Using a register asm variable seems to be causing lots of bugs.

Reply:
Does `__builtin_frame_address` work though, and is that sufficient for real-world uses?  Similarly, I wonder if PCC is  better served by `__builtin_return_address`?

Reply:
`__builtin_return_address` will get you the wrong bounds/permissions in some cases. `__builtin_frame_address` works as far as I know.

Reply:
__builtin_frame_address() does seem to work, so we could drop this.

__builtin_return address is not the same as cheri_getpcc(): the return address will be a sentry, whereas pcc is unsealed.

Reply:
What places will actually use `cheri_getpcc` in practice?  We might also need to think about making `__builtin_return_address` sufficiently functional if having it be a sentry isn't for current use cases.   I think FreeBSD uses it to get the starting point of a stack trace for the current thread in the kernel on various architectures, but all it really needs for that is to be able to read the virtual address.

Reply:
Having it not be a sentry is impossible. If you are able to get a non-sentry then the security argument vanishes.

Reply:
(But that doesn't mean a sentry is useless. You can use it for unwinding, just don't expect it to be tagged and dereferenceable, so if you need to poke at instructions, derive a suitable capability from elsewhere, but if you just need an address then it's sufficient.)

Reply:
RTLD and csu code uses cheri_getpcc() to create initial code pointers.

Reply:
I think we should drop this. People already use `__builtin_frame_address` when they mean that, we don't need another variant (that's also presumably wrong for hybrid code).